### PR TITLE
[Feature]: Submit all blocks possible

### LINKED ISSUE #1590: [Feature Request]: Submit all blocks possible
## Context 

#1550 added tests to the block submission that highlighted the following edge case.

Given a set of blocks, if any of the blocks has a marshaling error, none of the previous blocks would be submitted. 

For example, given the following test case:
```
name: "A and B are fair blocks, but C has a marshalling error. So C never gets submitted",
			blocks: func() []*types.Block {
				numBlocks, numTxs := 3, 5
				blocks := make([]*types.Block, numBlocks)
				blocks[0] = types.GetRandomBlock(uint64(1), numTxs)
				blocks[1] = types.GetRandomBlock(uint64(2), numTxs)
				blocks[2] = types.GetRandomBlock(uint64(3), numTxs)
				invalidateBlockHeader(blocks[2])
				return blocks
			}(),
			isErrExpected:               true,
			expectedPendingBlocksLength: 1,
```
This is the ideal end result, A and B would be submitted, while C isn't due to it's marshaling error. 
However, currently no blocks will be submitted because a marshal error immediately returns an error as oppose to using a `break` like exceeding the blob size. 

REF: https://github.com/rollkit/rollkit/blob/f61fa09916be6d25bc87654bf09d4d0eab847a3c/da/da.go#L113

## Proposal 

There are a few things that I think can be cleaned up in the `SubmitBlocks` method. 

1. `submitted`
The `submitted` variable is not needed and is a confusing variable. https://github.com/rollkit/rollkit/blob/f61fa09916be6d25bc87654bf09d4d0eab847a3c/da/da.go#L109
We can simply use `len(blobs)` here https://github.com/rollkit/rollkit/blob/f61fa09916be6d25bc87654bf09d4d0eab847a3c/da/da.go#L128

2. Break on marshal error
I think we should `break` on marshal error so we can submit any good blocks ahead of the bad block https://github.com/rollkit/rollkit/blob/f61fa09916be6d25bc87654bf09d4d0eab847a3c/da/da.go#L113

The result of 1 and 2 would be something like this:
```go
	var ( 
            blobs [][]byte
	    blobSize uint64
            code StatusCode
            message string
	)
	for i := range blocks {
		blob, err := blocks[i].MarshalBinary()
		if err != nil {
                         code = StatusError
			message = fmt.Sprint("failed to serialize block", err)
			dac.Logger.Info(message)
			break
		}
		if blobSize+uint64(len(blob)) > maxBlobSize {
                         code = StatusError
			message = fmt.Sprint(ErrBlobSizeOverLimit.Error(), "blob size limit reached", "maxBlobSize", maxBlobSize, "index", i, "blobSize", blobSize, "len(blob)", len(blob))
			dac.Logger.Info(message)
			break
		}
		blobSize += uint64(len(blob))
		blobs = append(blobs, blob)
	}
	if len(blobs) == 0 {
		return ResultSubmitBlocks{
			BaseResult: BaseResult{
				Code:    code,
				Message: "failed to submit blocks:  no blobs generated" + message,
			},
		}
	}
```

3. Improve handling of marshal error and blob limit exceeded with submit errors. 
In the case where `len(blobs) > 0` any errors or issues in the blob generation is forgotten about. We should either properly propagate those errors after any error handling of `dac.da.Submit`, or we should have a comment explaining that since we break out of the blob generating loop on the first error, that block will be cause the same error the next submission resulting in the error propagating at that point because `len(blobs)` will be zero the next time around.

### PR DESCRIPTION
**Pull request by ThanhNhann**
## Overview
closes #1590 

- This pr updates the logic of `DAClient.SubmitBlocks` so now it can submit all blocks possible
## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved error handling and messaging for block submission failures due to serialization issues or size limits in the `SubmitBlocks` method.
- **Tests**
	- Updated test cases `Test_submitBlocksToDA_BlockMarshalErrorCase1` and `Test_submitBlocksToDA_BlockMarshalErrorCase2` to reflect changes in block submission and error messaging.

### CODE REVIEW

File: da/da.go [hunk 1]
```diff
@@ -186,7 +185,7 @@ func (dac *DAClient) SubmitBlocks(ctx context.Context, blocks []*types.Block, ma
 		BaseResult: BaseResult{
 			Code:           StatusSuccess,
 			DAHeight:       binary.LittleEndian.Uint64(ids[0]),
-			SubmittedCount: submitted,
+			SubmittedCount: uint64(len(blobs)),
```
Comment (MSevey) [inline da/da.go:188]: I think this should be `len(ids)`.

**Review by MSevey — CHANGES_REQUESTED** [review]
Thanks for the PR!
Some minor edits.

File: da/da.go (hunk 1 above)
Comment (ThanhNhann) [reply da/da.go:188]: yeah you r right, Im updating

**Comment by ThanhNhann** [comment]
hi @MSevey, can you plz re-run the CI, I think the pr look good now

**Comment by MSevey** [comment]
> hi @MSevey, can you plz re-run the CI, I think the pr look good now
> 
> 

@ThanhNhann the CI ran but it failed. Can you look at the CI logs and see what the failure was?

**Comment by MSevey** [comment]
> > hi @MSevey, can you plz re-run the CI, I think the pr look good now
> 
> @ThanhNhann the CI ran but it failed. Can you look at the CI logs and see what the failure was?

```
=== RUN   TestPendingBlocks
    loggers.go:42: Test logs can be found: /tmp/TestPendingBlocks.log1017824086
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1dc3366]

goroutine 17141 [running]:
github.com/rollkit/rollkit/da.(*DAClient).SubmitBlocks(0xc0278a4900, {0x2988288, 0xc027895590}, {0xc0278f7bc0, 0x3, 0x0?}, 0x2800, 0x4511c8?)
	/home/runner/work/rollkit/rollkit/da/da.go:150 +0x726
github.com/rollkit/rollkit/block.(*Manager).submitBlocksToDA(0xc035d70780, {0x2988288, 0xc027895590})
	/home/runner/work/rollkit/rollkit/block/manager.go:901 +0x485
github.com/rollkit/rollkit/block.(*Manager).BlockSubmissionLoop(0xc035d70780, {0x2988288, 0xc027895590})
	/home/runner/work/rollkit/rollkit/block/manager.go:390 +0x1ec
github.com/rollkit/rollkit/node.(*FullNode).OnStart.func2()
	/home/runner/work/rollkit/rollkit/node/full.go:390 +0x89
github.com/rollkit/rollkit/types.(*ThreadManager).Go.func1()
	/home/runner/work/rollkit/rollkit/types/threadmanager.go:26 +0x87
created by github.com/rollkit/rollkit/types.(*ThreadManager).Go in goroutine 16913
	/home/runner/work/rollkit/rollkit/types/threadmanager.go:24 +0xe5
FAIL	github.com/rollkit/rollkit/node	51.883s
```

File: da/da.go [hunk 2]
```diff
@@ -119,35 +119,36 @@ func NewDAClient(da goDA.DA, gasPrice, gasMultiplier float64, ns goDA.Namespace,
...
+	if len(blobs) == 0 {
 		return ResultSubmitBlocks{
 			BaseResult: BaseResult{
 				Code:    StatusError,
-				Message: "failed to submit blocks: oversized block: " + ErrBlobSizeOverLimit.Error(),
+				Message: "failed to submit blocks: no blobs generated " + message,
 			},
 		}
 	}
+	dac.Logger.Info("Starting to submit blocks with %v blobs generated", len(blobs))
```
Comment (MSevey) [inline da/da.go:150]: to resolve the panic, I'd just remove this log line. 
that is the cleanest way, otherwise we have to add an ugly ctx catch. 
The shutdown is going to be caught by the context immediately in `dac.DA.Submit` anyways.

**Review by MSevey — APPROVED** [review]

**Review by gupadhyaya — APPROVED** [review]
